README.md
# Deck App Maker 
A small standalone bash script to automate making executable webapps for the Steam Deck platform. Similar utilities like WebcataloGNU General Public License v3.0g pack their own blends of Chromium or other browsers that restrict functionality such as extensions, leading the user to rely on their paid ad blocking solutions. This tool uses the default installation of Firefox and takes advantage of the --kiosk command line switch, allowing the user to browse with whichever extensions and tools they prefer.

## Installation
1. Place anywhere and execute from terminal. Deck App Maker uses default installation values and directories and will place the .desktop and icon files in ~/home/deck/.webapps *.
2. In Steam, go to "Add a Non-Steam Game to My Library" and navigate to /home/deck/.webapps and select the app. It should now show up in your library

*Other browsers and directories may be used but the script will need to be edited to reflect the desired file paths.

# License
GNU General Public License v3.0


changelog.md
# Changelog
Notable changes to project will be documented here

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- functionality for other browsers baked in
- a better solution for scraping favicons

## [1.0.1] - 2022-08-17
### Fixed
- fixe a bug that would break the script if app names with spaces were used
## [1.1.0] - 2022-08-18
- fixed an issue where steam would not execute a desktop app if the exec line had spaces. Deck App Maker now generates a bash script that executes the game and the .desktop file points to that.
- fixed some typos in the desktop file generation
- looped script instead of calling itself, will prevent loop from breaking when new version saved


deckappmaker.sh
#!/bin/bash
clear

#User inputs and var assignment
echo "DeckApp Maker"
echo -e "\n\n"
read -p "What would you like to call the app: " appName
read -p "Enter full URL for site you would like to create an app: " desiredPage

#simple script to scrape the desired page for favicons - will fail more often than succeeds
wget -P /home/deck/.webapps/$appName/data -nd -H -p -A  .ico,favicon*.png,ico.png,*favicon*.ico  -e robots=off $desiredPage

#saves largest and generally highest quality icon image and sets name as variable
find /home/deck/.webapps/$appName/data -type f | sort -n -r | tail -n +2 | xargs -I{} rm -v {}
icon=$(ls /home/deck/.webapps/$appName/data)

#generates desktop file that can be executed by steam
echo [Desktop Entry] > /home/deck/.webapps/$appName.desktop
echo Encoding=UTF-8 >> /home/deck/.webapps/$appName.desktop
echo Version=1.0 >> /home/deck/.webapps/$appName.desktop
echo Type=Application >> /home/deck/.webapps/$appName.desktop
echo Terminal=false >> /home/deck/.webapps/$appName.desktop
echo Exec=flatpak run org.mozilla.firefox --user-agent="\"Mozilla/5.0 (Android 13; Mobile; rv:68.0) Gecko/68.0 Firefox/103.0\"" --kiosk "$desiredPage" >> /home/deck/.webapps/$appName.desktop
echo Name=$appName >> /home/deck/.webapps/$appName.desktop
echo Icon=/home/deck/.webapps/$appName/data/$icon >> /home/deck/.webapps/$appName.desktop

#makes generated desktop file executable
chmod +x /home/deck/.webapps/$appName.desktop

echo -e "\n\n\n"
read -p "$appName was created at /home/deck/.webapps...Press any key to continue"
./deckappmaker.sh



src/deprecated/deckappmaker_1.0.1.sh
#!/bin/bash
clear

#User inputs and var assignment
echo "DeckApp Maker"
echo -e "\n\n"
read -p "What would you like to call the app: " appName
read -p "Enter full URL for site you would like to create an app: " desiredPage

#simple script to scrape the desired page for favicons - will fail more often than succeeds
wget -P /home/deck/.webapps/${appName// /_}/data -nd -H -p -A  .ico,favicon*.png,ico.png,*favicon*.ico  -e robots=off $desiredPage

#saves largest and generally highest quality icon image and sets name as variable
find /home/deck/.webapps/${appName// /_}/data -type f | sort -n -r | tail -n +2 | xargs -I{} rm -v {}
icon=$(ls /home/deck/.webapps/${appName// /_}/data)

#generates desktop file that can be executed by steam
echo [Desktop Entry] > /home/deck/.webapps/${appName// /_}.desktop
echo Encoding=UTF-8 >> /home/deck/.webapps/${appName// /_}.desktop
echo Version=1.0 >> /home/deck/.webapps/${appName// /_}.desktop
echo Type=Application >> /home/deck/.webapps/${appName// /_}.desktop
echo Terminal=false >> /home/deck/.webapps/${appName// /_}.desktop
echo Exec=flatpak run org.mozilla.firefox --user-agent="\"Mozilla/5.0 (Android 13; Mobile; rv:68.0) Gecko/68.0 Firefox/103.0\"" --kiosk "$desiredPage" >> /home/deck/.webapps/${appName// /_}.desktop
echo Name=$appName >> /home/deck/.webapps/${appName// /_}.desktop
echo Icon=/home/deck/.webapps/${appName// /_}/data/$icon >> /home/deck/.webapps/${appName// /_}.desktop

#makes generated desktop file executable
chmod +x /home/deck/.webapps/${appName// /_}.desktop

echo -e "\n\n\n"
read -p "$appName was created at /home/deck/.webapps...Press any key to continue"
./deckappmaker.sh



src/deckappmaker_1.1.0.sh
#!/bin/bash

while :
do

    clear

    #User inputs and var assignment
    echo "DeckApp Maker"
    echo -e "\n\n"
    read -p "What would you like to call the app: " appName
    read -p "Enter full URL for site you would like to create an app: " desiredPage

    #simple script to scrape the desired page for favicons - will fail more often than succeeds
    wget -P /home/deck/.webapps/${appName// /_}/data -nd -H -p -A  .ico,favicon*.png,ico.png,*favicon*.ico  -e robots=off $desiredPage

    #saves largest and generally highest quality icon image and sets name as variable
    find /home/deck/.webapps/${appName// /_}/data -type f | sort -n -r | tail -n +2 | xargs -I{} rm -v {}
    icon=$(ls /home/deck/.webapps/${appName// /_}/data)

    #generates bash file that can be executed by steam
    echo \#!/bin/bash >> /home/deck/.webapps/${appName// /_}/launcher.sh
    echo flatpak run org.mozilla.firefox --user-agent="\"Mozilla/5.0 (Android 13; Mobile; rv:68.0) Gecko/68.0 Firefox/103.0\"" --kiosk "$desiredPage" >> /home/deck/.webapps/${appName// /_}/launcher.sh

    #generates desktop file for steam library archive
    echo [Desktop Entry] >> /home/deck/.webapps/${appName// /_}/${appName// /_}.desktop
    echo Name=$appName >> /home/deck/.webapps/${appName// /_}/${appName// /_}.desktop
    echo Comment=Browse a Steam Deck friendly version of this website. >> /home/deck/.webapps/${appName// /_}/${appName// /_}.desktop
    echo Exec=/home/deck/.webapps/${appName// /_}/launcher.sh >> /home/deck/.webapps/${appName// /_}/${appName// /_}.desktop
    echo Icon=/home/deck/.webapps/${appName// /_}/data/$icon >> /home/deck/.webapps/${appName// /_}/{appName// /_}.desktop
    echo Terminal=false >> /home/deck/.webapps/${appName// /_}/${appName// /_}.desktop
    echo Type=Application >> /home/deck/.webapps/${appName// /_}/${appName// /_}.desktop
    echo Categories=App >> /home/deck/.webapps/${appName// /_}/${appName// /_}.desktop

    #makes generated desktop file executable
    chmod +x /home/deck/.webapps/${appName// /_}/${appName// /_}.desktop
    chmod +x /home/deck/.webapps/${appName// /_}/launcher.sh

    echo -e "\n\n\n"
    read -p "$appName was created at /home/deck/.webapps/${appName// /_}/..."
    echo -e "\n\n"

done
